rfb/rfbclient.h

```diff
@@ -477,6 +484,14 @@ typedef struct _rfbClient {
          * Used for intended dimensions, rfbClient.width and rfbClient.height are used to manage the real framebuffer dimensions.
 	 */
 	rfbExtDesktopScreen screen;
+
+#ifdef LIBVNCSERVER_HAVE_LIBZ
+	/**
+	 * Used for extended clipboard text fallback
+	 * When server announce cap of utf8, but still send latin-1
+	 */
+	GotXCutTextProc GotXCutTextFallback;
```

Comment:
If I understand correctly, encoding of text passed to `GotXCutText` callback will now depend on whether extended clipboard is enabled. While `GotXCutTextFallback` will always be called with Latin-1.

Now, `GotXCutText` callback implementations will always have to check `client->appData.clipboardEnabledCap` before parsing the data.

Would it not make sense to always call `GotXCutText` with Latin-1, irrespective of how Latin-1 text is received, and replace `GotXCutTextFallback` with `GotXCutTextUTF8` dedicated to UTF-8 encoded text.

This way you don't have to implement fallback handling here, and callback implementations become simpler.

Reply:
Thinking more about this, it is likely that support for data types other than simple text will be added in future. So the new callback should be something like: `GotXCutTextExt(int type, char* buffer, size_t buffer_len)`

Otherwise, separate callbacks will be needed for different data types.

Reply:
We here do mean one type data  throught one callback.
In future, for example, we want support file, then we add a callback for file.
The reason we use old GotCutText for both utf8 and latin-1 is that we had history and we want the existing vnc-client and this library cost little.

Reply:
@gujjwal00 GotXCutTextFallback can ignored for most cases, it is used to indicate some strange behavior, people can ignore it. It is different from a server that dont support utf8-text. Maybe it can used as a DEBUG tool to check server implementation.

Reply:
@gujjwal00 , In first path, there is no GotXCutTextFallback logical, we just log and ignore it. I think people can ignore it as usual :). I add this thing for people who say it is wanted...

Reply:
>We here do mean one type data throught one callback.
In future, for example, we want support file, then we add a callback for file.

Right now there are 4 data types defined (text, rtf, html, dib). I don't think LibVNCClient is ever going to parse any of these. It will most likely hand over the buffer to application, which will use platform-specific APIs to put them on clipboard. You have already done the hard part of handling Zlib streams here. After that apps will likely want to take over the decompressed buffer. Same thing applies when data is sent to the server.

File data type is only reserved, not defined yet. But you are right, it may very well require a dedicated callback.

> The reason we use old GotCutText for both utf8 and latin-1 is that we had history and we want the existing vnc-client and this library cost little.

Existing clients are not going to get UTF-8 support automatically. They will need to enable it in LibVNCClient and modify  `GotCutText` to handle both encodings. So work for them is same regardless. Similarly, you are adding a new callback for fallback handling anyway.

IMHO, using separate callback will be much cleaner, and LibVNCServer already uses similar approach. But this is just a suggestion.

>  In first patch, there is no GotXCutTextFallback logic, we just log and ignore it. I think people can ignore it as usual :)

I want to believe you, but the number of small implementation differences among VNC servers/clients makes me skeptical  😀.

Reply:
@gujjwal00 Some impl had have latin1/utf8 handler, they set it according to configuration. This implementation rely on configuration, old code have set callback for utf8 if configured(but want not work) and they can set rfbclient.appData.clipboardCap = rfbExtendedClipboard_Text to let it work.
uft8 or other encoding's process signature is the same, that is why some implemenation is ready for utf8 if they support this configuation.

Reply:
@gujjwal00 if client and server process handshake correctly and both agree on clipboard utf8, client(it is us) always send in utf8, and the server sometime send latin1 in old way, I think there maybe some BUG or report it.
When we discuss, we dont talk about the sending-path, current design dont give new sending method, if both agree on utf8, client had to send utf8.
As I know, libvncserver will always send utf8 if both agreed and client always send utf8.
Tigervnc is my main testbed, it always send utf8 if both agree on utf8 and client always send utf8.
Maybe more server will be tested.

But if both agree on utf8, and client always send utf8, how server can change its state to send in latin1?

Back to api interface, we choose not to add new api for both sending and receiving. For most case, you can use the old api without bothering GotXCutTextFallback.

Reply:
@gujjwal00 Based on the fact that utf8/latin1 both use the same message, and we simplify/limit the client's behavior that if both agree on utf8, it just can send in new way(field length < 0); and if not agree on utf8, it just can send text in old way(field length >=0). Current choice of API may be better.
If we device SendUtf8/GotUtf8 API, it make thing more complex.

Reply:
@wuhanck @gujjwal00 I took the freedom to rework the API to resemble the LibVNCServer one in fed2c8b15f1bf31251691da588cb5fb6c37b262b. This PR got merged as c0b3601dfe8212a3683750b1e0dfb7f343e2ab83.